proof/refine/ARM/Schedule_R.thy

```diff
@@ -3368,5 +3317,840 @@ lemma possible_switch_to_corres:
+         apply (rule_tac s'=s' in refills_heads_equal_active[THEN conjunct2, symmetric])
+           apply simp
+          apply blast
+         apply blast
+        apply wpsimp+
+   apply (clarsimp simp: obj_at_def is_sc_obj_def)
+  apply (fastforce dest: sc_ko_at_valid_objs_valid_sc')
+  done
+
+lemma getTCBRefillReady_corres:
+  "corres (=) (pspace_aligned and pspace_distinct and valid_objs and active_sc_tcb_at t) valid_objs'
+          (get_tcb_refill_ready t)
+          (getTCBRefillReady t)"
+   (is "corres _ ?abs ?conc _ _")
+  apply (rule corres_cross[where Q'="tcb_at' t", OF tcb_at'_cross_rel])
+   apply (fastforce simp: vs_all_heap_simps obj_at_def is_tcb_def)
+  apply (clarsimp simp: get_tcb_refill_ready_def get_tcb_obj_ref_def getTCBRefillReady_def)
+  apply (rule corres_guard_imp)
+    apply (rule_tac Q="?abs" and Q'="?conc"
+                 in corres_split[OF _ threadget_corres[where r="(=)"] thread_get_sp threadGet_inv])
+       apply (rename_tac sc sc')
+       defer
+     apply (clarsimp simp: tcb_relation_def)
+    apply (clarsimp simp: vs_all_heap_simps obj_at_def is_tcb_def)
+   apply simp
+  apply (rule corres_symb_exec_l)
+     apply (rename_tac sc_ptr)
+     apply (rule_tac F="sc_ptr = the sc'" in corres_gen_asm)
+     apply (rule corres_guard_imp)
+       apply (rule refillReady_corres)
+       apply blast
+      apply simp
+     apply simp
+    apply (clarsimp simp: exs_valid_def obj_at_def return_def vs_all_heap_simps)
+   apply wpsimp
+   apply (fastforce intro: valid_sched_context_size_objsI
+                     simp: obj_at_def vs_all_heap_simps is_sc_obj_def)
+  apply (clarsimp simp: no_fail_def obj_at_def return_def vs_all_heap_simps)
+  done
+
+lemma getReleaseQueue_corres:
+  "corres (=) \<top> \<top> (gets release_queue) (getReleaseQueue)"
+  unfolding getReleaseQueue_def
+  apply (rule corres_gets_trivial)
+  by (clarsimp simp: state_relation_def release_queue_relation_def)
+
+lemma releaseQNonEmptyAndReady_corres:
+  "corres (=) (pspace_aligned and pspace_distinct and valid_objs and valid_release_q) valid_objs'
+              release_q_non_empty_and_ready releaseQNonEmptyAndReady"
+  (is "corres _ ?abs _ _ _")
+  apply (clarsimp simp: release_q_non_empty_and_ready_def releaseQNonEmptyAndReady_def)
+  apply (rule corres_guard_imp)
+    apply (rule corres_split[OF _ getReleaseQueue_corres])
+      apply clarsimp
+      apply (rename_tac rq)
+      apply (rule_tac P="?abs and (\<lambda>s. release_queue s = rq
+                                       \<and> (release_queue s \<noteq> []
+                                          \<longrightarrow> active_sc_tcb_at (hd (release_queue s)) s))"
+                   in corres_inst)
+      apply (rule corres_guard_imp)
+        apply (rule getTCBRefillReady_corres)
+       apply fastforce
+      apply simp
+     apply wpsimp
+    apply wpsimp
+   apply (clarsimp simp: valid_release_q_def)
+  apply simp
+  done
+
+(* FIXME RT: move? NonDetMonad? *)
```

Comment:
Yes, this should go into NonDetMonad (together with the corresponding lemmas, some which might be fine in NonDetMonadLemmas)